ports/stm32/dma.c

```diff
@@ -448,10 +448,17 @@ void dma_init(DMA_HandleTypeDef *dma, const dma_descr_t *dma_descr, void *data){
         } else {
             // only necessary initialization
             dma->State = HAL_DMA_STATE_READY;
+
 #if defined(MCU_SERIES_F4) || defined(MCU_SERIES_F7)
             // calculate DMA base address and bitshift to be used in IRQ handler
             extern uint32_t DMA_CalcBaseAndBitshift(DMA_HandleTypeDef *hdma);
             DMA_CalcBaseAndBitshift(dma);
+
+#elif defined(MCU_SERIES_L4)
+            // The memset() just above will have cleared these private vars from the HAL.
+            // Recalculate them now.
+            dma->DmaBaseAddress = (DMA_TypeDef *)((uint32_t)dma->Instance & ~0xff);
+            dma->ChannelIndex = (((uint32_t)dma->Instance - (uint32_t)dma->DmaBaseAddress) /20)*4;
```

Comment:
I'm not sure this is correct: dma->DmaBaseAddress is not the same as DMA1_Channel1, is it?

Reply:
`DmaBaseAddress` is either `DMA1` or `DMA2`. `DMA1_Channel1` has an offset corresponding to which channel it is. The above masking and such can only be understood by looking at the DMA register map.

Reply:
Ooops, this is probably due to my L4 port... 
On L4 the dma->ChanelIndex can be obtained from the expression ((dma_descr->id)%NSTREAMS_PER_CONTROLLER)<<2;
and the DMABaseadresse could be obtained by comparing the dma_descr->id  with NSTREAMS_PER_CONTROLLER. 
I would propose to modify the dma_init_handle to encapsulate any setup of the handle in the existing dma?init_handle function in the following way:
```C
void dma_init_handle(DMA_HandleTypeDef *dma, const dma_descr_t *dma_descr, void *data) {
    // initialise parameters
    dma->Instance = dma_descr->instance;
    dma->Init = *dma_descr->init;
    dma->Init.Direction = dma_descr->transfer_direction;
    #if defined(MCU_SERIES_L4)
    dma->Init.Request = dma_descr->sub_instance;
    // Factor 4 (or <<2) for matching channel bit location with
    // CSELR, IFCR, ISR .. register.
    // See as well HAL_DMA_init()
    dma->ChannelIndex = ((dma_descr->id)%NSTREAMS_PER_CONTROLLER)<<2;
    if (dma_descr->id < NSTREAMS_PER_CONTROLLER) {
    	dma->DmaBaseAddress = DMA1;
    } else {
    	dma->DmaBaseAddress = DMA2;
    }
    #else
    dma->Init.Channel = dma_descr->sub_instance;
    // calculate DMA base address and bitshift to be used in IRQ handler
    extern uint32_t DMA_CalcBaseAndBitshift(DMA_HandleTypeDef *hdma);
    DMA_CalcBaseAndBitshift(dma);
    #endif
    // half of __HAL_LINKDMA(data, xxx, *dma)
    // caller must implement other half by doing: data->xxx = dma
    dma->Parent = data;
}
```
and remove any #defines in the dma_init routine. I will test it on my DMA for UART branch and will give feedback here.